### ISSUE TITLE
query shows bogus results

### ISSUE THREAD

**Issue by auhlig** [prometheus/prometheus#3776]
**What did you do?**
We're using 2 Prometheis in a federation. A "collector" and a "frontend". The collector gets the metrics from, in this case, cAdvisor. The frontend federates aggregated data from the collector ([config](https://github.com/sapcc/helm-charts/blob/master/system/kube-monitoring/charts/prometheus-frontend/templates/config.yaml)). While querying for `container_cpu_system_seconds_total` in the collector returns the expected results, the same query in the frontend shows other metrics. See screenshot below.

**What did you expect to see?**
only `container_cpu_system_seconds_total` metrics

**What did you see instead? Under which circumstances?**
As shown in the screenshot below, the query for `container_cpu_system_seconds_total` also shows unrelated metrics.
![query](https://user-images.githubusercontent.com/14873188/35671508-e6ead720-073b-11e8-8e35-d25e6da278ab.png)

**Environment**

* System information:

Linux 4.14.11-coreos x86_64

* Prometheus version:

prometheus, version 2.1.0 (branch: HEAD, revision: 85f23d82a045d103ea7f3c89a91fba4a93e6367a)
  build user:       root@6e784304d3ff
  build date:       20180119-12:01:23
  go version:       go1.9.2

* Prometheus configuration file:

Can be found [here](https://github.com/sapcc/helm-charts/blob/master/system/kube-monitoring/charts/prometheus-frontend/templates/config.yaml).

**Comment by brian-brazil** [comment]
That's not good. What versions of Prometheus has been run using the same storage?

**Comment by brian-brazil** [comment]
Also, incidentally you should read https://www.robustperception.io/federation-what-is-it-good-for/

**Comment by RichiH** [comment]
Unlikely, but do you have a proxy or cache in front of the Prometheus instance you're querying?

**Comment by auhlig** [comment]
The same storage was used by the previous version, Prom 2.0.0 . We have this setting in 16 regions, however only 1 showed the above bogus result. No proxy or cache involved.

**Comment by brian-brazil** [comment]
Are there any indications of disk data corruption on the machine? We have checksums so should catch that, but you might have gotten really unlucky.

**Comment by auhlig** [comment]
There were issues in the past related to the retention fix https://github.com/prometheus/prometheus/issues/3534. However Prometheus 2.0.0 showed the correct results.

**Comment by fabxc** [comment]
@auhlig sorry about the delay on this as it is quite severe.

If you roll back to Prometheus 2.0.0, do the results become correct again. How big are your `index` files in the storage directory?

**Comment by auhlig** [comment]
Hi @fabxc,
Thanks for the reply. 
Since this issue was quite visible, as it messed up our Grafana dashboards, we had to wipe the data in the affected prometheus as a quick fix. Thus I cannot test this anymore. The fresh prometheus works as expected.

**Comment by brian-brazil** [comment]
Without the data, we can only presume that this fixed unless someone else reports it.

**Comment by auhlig** [comment]
Another instance: Query for `go_goroutines` shows also `go_memstats_*`.
 ![prom_foo](https://user-images.githubusercontent.com/14873188/37780003-11eef7d4-2dee-11e8-9bc0-f7fae94f5149.png)

I can see about a trillion `*.tmp` folders containing only data (no index, meta.json)

**Comment by fabxc** [comment]
Nothing in the logs? Left over .tmp folders would usually indicate some kind of failed compactions.

**Comment by fabxc** [comment]
Can you provide us one or more blocks for affected time ranges before wiping this time?

**Comment by auhlig** [comment]
Grepping for `component=tsdb` shows 3 types of errors:

(1) ```level=error ts=2018-03-22T15:35:09.928277674Z caller=db.go:265 component=tsdb msg="compaction failed" err="persist head block: write compaction: add series: out-of-order series added with label set ...```
(2) ```level=info ts=2018-03-22T15:35:04.79684528Z caller=compact.go:387 component=tsdb msg="compact blocks" count=1 mint=1520755200000 maxt=1520762400000```
(3) ```level=error ts=2018-03-22T15:35:09.928191279Z caller=compact.go:397 component=tsdb msg="removed tmp folder after failed compaction" err="remove /prometheus/01C975JSTWD7EGP3GPM1S005MV.tmp/chunks: directory not empty"```

**Comment by auhlig** [comment]
Something like the following?
```
/prometheus # du -ha 01C6SMSMDYM5M8GVSQG8X6J50S.tmp
168.7M	01C6SMSMDYM5M8GVSQG8X6J50S.tmp/chunks/000001
168.7M	01C6SMSMDYM5M8GVSQG8X6J50S.tmp/chunks
32.1M	01C6SMSMDYM5M8GVSQG8X6J50S.tmp/index
200.8M	01C6SMSMDYM5M8GVSQG8X6J50S.tmp
```
Most of the *.tmp folders look like:
```
prometheus # du -ha 01C87J6S6RNDN9ZWSYJGB34PNC.tmp
4.0K	01C87J6S6RNDN9ZWSYJGB34PNC.tmp/chunks
8.0K	01C87J6S6RNDN9ZWSYJGB34PNC.tmp

```

**Comment by fabxc** [comment]
The `.tmp` folders are just leftovers from failed compactions. We don't really care about them. From the logs I take that it actually never manages to write a block with the `.tmp` suffixes? This definitely means this is not a disk corruption issue.

That's pretty unfortunate for remote debugging. In the `(1)` log line you are showing, is it always the same series shown where you subtituted for with `...` or do they differ? Full examples would help.

Log message (3) confuses me as it apparently cannot remove the chunk directory. I thought it was an issue in the code but we are actually [calling `os.RemoveAll`](https://github.com/prometheus/tsdb/blob/master/compact.go#L420) which should delete directories recursively without issues.

What Prometheus version are you using by now? Note that 2.1.0 and 2.2.0 both have issues and you should definitely upgrade to 2.2.1.

**Comment by auhlig** [comment]
Regarding the blocks:
```
/prometheus # find . -name index -type f -exec ls -lh {} \;
-rw-r--r--    1 root     root       32.0M Feb 20 13:00 ./01C6SMSMDYM5M8GVSQG8X6J50S.tmp/index
-rw-r--r--    1 root     root        4.5M Mar 11 09:00 ./01C8A4KG1KBVYG3X7N41WPXDAY/index
-rw-r--r--    1 root     root       40.7M Mar  9 13:00 ./01C85DHQJVMJ4WSBKXWKCBBG5D/index
-rw-r--r--    1 root     root       44.3M Mar  7 19:00 ./01C80XBF9P97ZH113END42ZZ3H/index
-rw-r--r--    1 root     root       40.2M Mar  8 01:00 ./01C81HYN45GJHS40YC62NWV7SJ/index
-rw-r--r--    1 root     root       40.2M Mar  8 07:00 ./01C826HV2ARSMRSST9M72VFH1P/index
-rw-r--r--    1 root     root       40.3M Mar  8 13:00 ./01C82V50X7EHK065M28QW1B2WA/index
-rw-r--r--    1 root     root       42.0M Mar  8 19:00 ./01C83FR6JY4140V5XE92EAGXZG/index
-rw-r--r--    1 root     root       40.3M Mar  9 01:00 ./01C844BC57HF5GSGM9Z8YVK0T4/index
-rw-r--r--    1 root     root       40.5M Mar  9 07:00 ./01C84RYJ33CRVY5TEDMSAMFPXF/index
-rw-r--r--    1 root     root       40.5M Mar  9 19:00 ./01C8624X5EB1YQE4JPRP6QVSG0/index
-rw-r--r--    1 root     root       40.6M Mar 10 01:00 ./01C86PR3CMXAVJ7D23CWQRYHWG/index
-rw-r--r--    1 root     root       40.7M Mar 10 07:00 ./01C87BB95Q3CZ98P1QRVSE1YHB/index
-rw-r--r--    1 root     root       44.2M Mar 11 01:00 ./01C8994QGGH9J01P09B6QMX0YN/index
-rw-r--r--    1 root     root       44.1M Mar 11 07:00 ./01C89XQXM877C5HHJTEV5WCT60/index
-rw-r--r--    1 root     root       39.7M Mar  4 13:00 ./01C7RHJ440HKK4KFJKR5ZYRS31/index
-rw-r--r--    1 root     root       39.8M Mar  4 19:00 ./01C7S65AMP0MSH22FJSKMBNFWK/index
-rw-r--r--    1 root     root       39.7M Mar  5 01:00 ./01C7STRG4BVHQ80EQPP8NSPHDR/index
-rw-r--r--    1 root     root       40.0M Mar  5 07:00 ./01C7TFBNCQ7WSG57XE0WR876V1/index
-rw-r--r--    1 root     root       41.5M Mar  5 13:00 ./01C7V3YVMQ0H2X8QRTTTPRY053/index
-rw-r--r--    1 root     root       40.6M Mar  5 19:00 ./01C7VRJ1CFFN7MMDW4G9RT9747/index
-rw-r--r--    1 root     root       40.6M Mar  7 13:00 ./01C808R9VNT5ASJZKXF1B35RC5/index
-rw-r--r--    1 root     root       40.3M Mar  6 01:00 ./01C7WD573G8QFZEJ1JEW7NHV2X/index
-rw-r--r--    1 root     root       40.2M Mar  6 07:00 ./01C7X1RCQYXGKS2THYG2PH51BG/index
-rw-r--r--    1 root     root       40.9M Mar  6 13:00 ./01C7XPBJHTKK4MTVQMZDZFBEMN/index
-rw-r--r--    1 root     root       40.4M Mar  6 19:00 ./01C7YAYRT12W1FR3J6EEBB97DT/index
-rw-r--r--    1 root     root       40.1M Mar  7 01:00 ./01C7YZHY33A4EMYTTZB82B7CWS/index
-rw-r--r--    1 root     root       40.1M Mar  7 07:00 ./01C7ZM53JTX2YCGNY5DCRCYYQJ/index
-rw-r--r--    1 root     root       43.2M Mar 10 19:00 ./01C88MHHGTPC9XP72TQVVKCS6V/index
-rw-r--r--    1 root     root       20.0M Mar 10 13:00 ./01C87ZYERSPXWZ5350Y9NZYHGG.tmp/index
-rw-r--r--    1 root     root       49.8M Mar 10 13:00 ./01C87ZYC6R6KRTSHM3ER1GNNGN/index
```

Regarding (1): It's always the same series as shown below.
```
level=error ts=2018-03-22T16:05:13.960244727Z caller=db.go:265 component=tsdb msg="compaction failed" err="persist head block: write compaction: add series: out-of-order series added with label set \"{__name__=\\\"apiserver_request_count\\\",client=\\\"Swagger-Codegen/1.0.0-alpha/python\\\",code=\\\"200\\\",component=\\\"apiserver\\\",contentType=\\\"application/json\\\",instance=\\\"<removedTheNodeName>\\\",job=\\\"kube-system/apiserver\\\",region=\\\"ap-cn-1\\\",resource=\\\"deployments\\\",verb=\\\"PATCH\\\"}\""
```

Regarding [os.RemoveAll](https://github.com/prometheus/tsdb/blob/master/compact.go#L420): Is the directory we're trying to delete used by another goroutine? That could potentially lead to the error.

We're still the same Prometheus version:
```
/prometheus # prometheus --version
prometheus, version 2.1.0 (branch: HEAD, revision: 85f23d82a045d103ea7f3c89a91fba4a93e6367a)
  build user:       root@6e784304d3ff
  build date:       20180119-12:01:23
  go version:       go1.9.2
```
I'll update staging today and continue with  our productive regions next week.

**Comment by brian-brazil** [comment]
We believe this was fixed in 2.3.0 as part of the big promql changes, if you still see this after that version please let us know.
